astropy/io/fits/fitstime.py

```diff
@@ -0,0 +1,251 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import re
+import warnings
+from collections import defaultdict, OrderedDict
+
+import numpy as np
+
+from . import Header, Card
+
+from ...coordinates import EarthLocation
+from ...table import Column
+from ...time import Time
+from ...utils.exceptions import AstropyUserWarning
+
+
+class FITS_time(object):
+    """A class to read FITS binary table time columns as `~astropy.time.Time`
+
+    This class reads the metadata associated with time coordinates, as 
+    stored in a FITS binary table header, converts time columns into 
+    `~astropy.time.Time` columns and reads global reference times as 
+    `~astropy.time.Time` instances.
+    """
+
+    # Global time reference coordinate keywords
+    TIME_KEYWORDS = {'TIMESYS' : 'scale',
+                     'MJDREF' : 'mjdref',
+                     'JDREF' : 'jdref',
+                     'DATEREF' : 'dateref',
+                     'TREFPOS' : 'pos',
+                     'TREFDIR' : 'dir',
+                     'TIMEUNIT' : 'unit',
+                     'TIMEOFFS' : 'offs',
+                     'OBSGEO-X' : 'loc_x',
+                     'OBSGEO-Y' : 'loc_y',
+                     'OBSGEO-Z' : 'loc_z',
+                     'DATE' : 'date',
+                     'DATE-OBS' : 'date_obs',
+                     'DATE-END' : 'date_end'}
+
+    # Column-specific time override keywords
+    COLUMN_TIME_KEYWORDS = {'TCTYP' : 'scale',
+                            'TRPOS' : 'pos',
+                            'TCUNI' : 'unit'}
+
+    # AstroPy FITS File specification
+    ASTROPY_SPECIFIER = ('+---AstroPy FITS File---+', 
+                         '/ Time information block')
+
+    # Set AstroPy Time global information
+    GLOBAL_TIME_INFO = {'TIMESYS' : ('UTC','Default time scale'),
+                        'JDREF' : (0.0,'Time columns are jd = jd1 + jd2'),
+                        'TREFPOS' : ('TOPOCENTER','Time reference position')}
+
+    # Compatibility of Time Scales and Reference Positions
+    TIME_SCALE_REF = {'tai': 'TOPOCENTER',
+                      'tt' : 'TOPOCENTER',
+                      'ut1' : 'TOPOCENTER',
+                      'utc' : 'TOPOCENTER',
+                      'tcg' : 'GEOCENTER',
+                      'tcb' : 'BARYCENTER',
+                      'tdb' : 'BARYCENTER'}
+
+    def __init__(self):
+        # Set defaults for global time scale, reference, etc.
+        self.global_info = {'scale' : 'UTC',
+                            'mjdref' : None,
+                            'unit' : 's'}
+        # Set default dictionary for time columns
+        self.time_columns = defaultdict(OrderedDict)
```

Comment:
I would have never thought of this construction, but it's really clever. Took me a while to understand what's going on here!